File: src/exoticatechnologies/modifications/exotics/impl/DaemonCore.java
package exoticatechnologies.modifications.exotics.impl;

import com.fs.starfarer.api.campaign.CampaignFleetAPI;
import com.fs.starfarer.api.campaign.econ.MarketAPI;
import com.fs.starfarer.api.combat.CombatEntityAPI;
import com.fs.starfarer.api.combat.DamageAPI;
import com.fs.starfarer.api.combat.MutableShipStatsAPI;
import com.fs.starfarer.api.combat.ShipAPI;
import com.fs.starfarer.api.combat.listeners.DamageDealtModifier;
import com.fs.starfarer.api.combat.listeners.DamageTakenModifier;
import com.fs.starfarer.api.fleet.FleetMemberAPI;
import com.fs.starfarer.api.ui.TooltipMakerAPI;
import com.fs.starfarer.api.ui.UIComponentAPI;
import com.fs.starfarer.api.util.Misc;
import exoticatechnologies.modifications.ShipModifications;
import exoticatechnologies.modifications.exotics.ExoticData;
import exoticatechnologies.util.StringUtils;
import exoticatechnologies.util.Utilities;
import org.json.JSONObject;
import org.lwjgl.util.vector.Vector2f;

import java.awt.*;
import java.util.HashMap;
import java.util.Map;

public class DaemonCore extends HullmodExotic {
    private static final String ITEM = "tahlan_daemoncore";

    public DaemonCore(String key, JSONObject settingsObj) {
        super(key, settingsObj, "et_daemoncore", "DaemonCore", new Color(150,20,20));
    }

    @Override
    public boolean shouldShow(FleetMemberAPI member, ShipModifications mods, MarketAPI market) {
        return canAfford(member.getFleetData().getFleet(), market)
                || Utilities.hasExoticChip(member.getFleetData().getFleet().getCargo(), this.getKey())
                || Utilities.hasExoticChip(Misc.getStorageCargo(market), this.getKey());
    }

    @Override
    public boolean canAfford(CampaignFleetAPI fleet, MarketAPI market) {
        return Utilities.hasItem(fleet.getCargo(), ITEM)
                || (Misc.getStorageCargo(market) != null && Utilities.hasItem(Misc.getStorageCargo(market), ITEM));
    }

    @Override
    public boolean removeItemsFromFleet(CampaignFleetAPI fleet, FleetMemberAPI fm, MarketAPI market) {
        if (Utilities.hasItem(fleet.getCargo(), ITEM)) {
            Utilities.takeItemQuantity(fleet.getCargo(), ITEM, 1);
        } else {
            Utilities.takeItemQuantity(Misc.getStorageCargo(market), ITEM, 1);
        }
        return true;
    }

    @Override
    public void modifyToolTip(TooltipMakerAPI tooltip, UIComponentAPI title, FleetMemberAPI fm, ShipModifications mods, boolean expand) {
        if (expand) {
            StringUtils.getTranslation(this.getKey(), "longDescription")
                    .addToTooltip(tooltip, title);
        }
    }

    @Override
    public Map<String, Float> getResourceCostMap(FleetMemberAPI fm, ShipModifications mods, MarketAPI market) {
        Map<String, Float> resourceCosts = new HashMap<>();
        resourceCosts.put(ITEM, 1f);
        return resourceCosts;
    }


    @Override
    public void applyExoticToStats(String id, FleetMemberAPI fm, MutableShipStatsAPI stats, ExoticData data) {
        onInstall(fm);
    }

    @Override
    public void applyExoticToShip(String id, FleetMemberAPI fm, ShipAPI ship, ExoticData data) {
        if (!ship.hasListenerOfClass(DaemonCoreDamageTakenListener.class)) {
            ship.addListener(new DaemonCoreDamageTakenListener());
        }

        if (!ship.hasListenerOfClass(DaemonCoreDamageDealtListener.class)) {
            ship.addListener(new DaemonCoreDamageDealtListener());
        }
    }

    /**
     * extra bandwidth added directly to ship.
     *
     * @param fm
     * @param mods
     * @param data
     * @return
     */
    public float getExtraBandwidth(FleetMemberAPI fm, ShipModifications mods, ExoticData data) {
        return 60f;
    }

    public class DaemonCoreDamageTakenListener implements DamageTakenModifier {
        @Override
        public String modifyDamageTaken(Object param, CombatEntityAPI target, DamageAPI damage, Vector2f point, boolean shieldHit) {
            damage.getModifier().modifyMult(DaemonCore.this.getBuffId(), 1.2f);
            return null;
        }
    }

    public class DaemonCoreDamageDealtListener implements DamageDealtModifier {
        @Override
        public String modifyDamageDealt(Object param, CombatEntityAPI target, DamageAPI damage, Vector2f point, boolean shieldHit) {
            damage.getModifier().modifyMult(DaemonCore.this.getBuffId(), 1.2f);
            return null;
        }
    }
}

File: src/exoticatechnologies/modifications/exotics/impl/DaemonCore.kt
package exoticatechnologies.modifications.exotics.impl

import com.fs.starfarer.api.Global
import com.fs.starfarer.api.campaign.CampaignFleetAPI
import com.fs.starfarer.api.campaign.econ.MarketAPI
import com.fs.starfarer.api.combat.CombatEntityAPI
import com.fs.starfarer.api.combat.DamageAPI
import com.fs.starfarer.api.combat.MutableShipStatsAPI
import com.fs.starfarer.api.combat.ShipAPI
import com.fs.starfarer.api.combat.listeners.DamageDealtModifier
import com.fs.starfarer.api.combat.listeners.DamageTakenModifier
import com.fs.starfarer.api.fleet.FleetMemberAPI
import com.fs.starfarer.api.ui.TooltipMakerAPI
import com.fs.starfarer.api.ui.UIComponentAPI
import com.fs.starfarer.api.util.Misc
import exoticatechnologies.modifications.ShipModifications
import exoticatechnologies.modifications.exotics.ExoticData
import exoticatechnologies.util.StringUtils
import exoticatechnologies.util.Utilities
import org.json.JSONObject
import org.lwjgl.util.vector.Vector2f
import java.awt.Color

class DaemonCore(key: String, settingsObj: JSONObject) :
    HullmodExotic(key, settingsObj, "et_daemoncore", "DaemonCore", Color(150, 20, 20)) {

    override fun getSalvageChance(chanceMult: Float): Float {
        return 0f
    }

    override fun shouldShow(member: FleetMemberAPI, mods: ShipModifications, market: MarketAPI?): Boolean {
        return (canAfford(Global.getSector().playerFleet, market)
                || Utilities.hasExoticChip(
            Global.getSector().playerFleet.cargo,
            key
        ) || Misc.getStorageCargo(market) != null && Utilities.hasExoticChip(Misc.getStorageCargo(market), key))
    }

    override fun canAfford(fleet: CampaignFleetAPI, market: MarketAPI?): Boolean {
        return Utilities.hasItem(
            fleet.cargo,
            ITEM
        ) || Misc.getStorageCargo(market) != null && Utilities.hasItem(Misc.getStorageCargo(market), ITEM)
    }

    override fun removeItemsFromFleet(fleet: CampaignFleetAPI, member: FleetMemberAPI, market: MarketAPI?): Boolean {
        if (Utilities.hasItem(fleet.cargo, ITEM)) {
            Utilities.takeItemQuantity(fleet.cargo, ITEM, 1f)
        } else {
            Utilities.takeItemQuantity(Misc.getStorageCargo(market), ITEM, 1f)
        }
        return true
    }

    override fun modifyToolTip(
        tooltip: TooltipMakerAPI,
        title: UIComponentAPI,
        member: FleetMemberAPI,
        mods: ShipModifications,
        exoticData: ExoticData,
        expand: Boolean
    ) {
        if (expand) {
            StringUtils.getTranslation(key, "longDescription")
                .format("bandwidthIncrease", BANDWIDTH_INCREASE * exoticData.type.getPositiveMult(member, mods))
                .format("smallReduction", SMALL_REDUCTION)
                .format("medReduction", MEDIUM_REDUCTION)
                .format("largeReduction", LARGE_REDUCTION)
                .format("fghtrReduction", FIGHTER_REDUCTION)
                .format("bmberReduction", BOMBER_REDUCTION)
                .format("doubleEdge", DOUBLE_EDGE * exoticData.type.getPositiveMult(member, mods))
                .addToTooltip(tooltip, title)
        }
    }

    override fun getResourceCostMap(
        fm: FleetMemberAPI,
        mods: ShipModifications,
        market: MarketAPI?
    ): MutableMap<String, Float> {
        val resourceCosts: MutableMap<String, Float> = HashMap()
        resourceCosts[ITEM] = 1f
        return resourceCosts
    }

    override fun applyExoticToStats(
        id: String,
        stats: MutableShipStatsAPI,
        member: FleetMemberAPI,
        mods: ShipModifications,
        exoticData: ExoticData
    ) {
        onInstall(member)
    }

    override fun applyToShip(
        id: String,
        member: FleetMemberAPI,
        ship: ShipAPI,
        mods: ShipModifications,
        exoticData: ExoticData
    ) {
        if (!ship.hasListenerOfClass(DaemonCoreDamageTakenListener::class.java)) {
            ship.addListener(DaemonCoreDamageTakenListener(member, mods, exoticData))
        }
        if (!ship.hasListenerOfClass(DaemonCoreDamageDealtListener::class.java)) {
            ship.addListener(DaemonCoreDamageDealtListener(member, mods, exoticData))
        }
    }

    /**
     * extra bandwidth added directly to ship.
     *
     * @param member
     * @param mods
     * @param exoticData
     * @return
     */
    override fun getExtraBandwidth(member: FleetMemberAPI, mods: ShipModifications, exoticData: ExoticData?): Float {
        return 60f * (exoticData?.type?.getPositiveMult(member, mods) ?: 1f)
    }

    inner class DaemonCoreDamageTakenListener(val member: FleetMemberAPI, val mods: ShipModifications, val exoticData: ExoticData) : DamageTakenModifier {
        override fun modifyDamageTaken(
            param: Any?,
            target: CombatEntityAPI,
            damage: DamageAPI,
            point: Vector2f,
            shieldHit: Boolean
        ): String {
            damage.modifier.modifyMult(buffId, 1 + (DOUBLE_EDGE / 100f) * exoticData.type.getPositiveMult(member, mods))
            return buffId
        }
    }

    inner class DaemonCoreDamageDealtListener(val member: FleetMemberAPI, val mods: ShipModifications, val exoticData: ExoticData) : DamageDealtModifier {
        override fun modifyDamageDealt(
            param: Any?,
            target: CombatEntityAPI,
            damage: DamageAPI,
            point: Vector2f,
            shieldHit: Boolean
        ): String {
            damage.modifier.modifyMult(buffId, 1 + (DOUBLE_EDGE / 100f) * exoticData.type.getPositiveMult(member, mods))
            return buffId
        }
    }

    companion object {
        private const val ITEM = "tahlan_archdaemoncore"
        private const val BANDWIDTH_INCREASE = 60
        private const val SMALL_REDUCTION = 1
        private const val MEDIUM_REDUCTION = 2
        private const val LARGE_REDUCTION = 4
        private const val FIGHTER_REDUCTION = 2
        private const val BOMBER_REDUCTION = 4
        private const val DOUBLE_EDGE = 20f
    }
}